File: apps/android/app/src/main/kotlin/com/nogeass/passalarm/domain/usecase/CreatePlanUseCase.kt
package com.nogeass.passalarm.domain.usecase

import com.nogeass.passalarm.domain.model.AlarmPlan
import com.nogeass.passalarm.domain.repository.AlarmPlanRepository
import javax.inject.Inject

class CreatePlanUseCase @Inject constructor(
    private val planRepository: AlarmPlanRepository,
    private val checkProLimit: CheckProLimitUseCase,
    private val reschedule: RescheduleNextNUseCase
) {
    suspend fun execute(plan: AlarmPlan): Boolean {
        if (!checkProLimit.canCreatePlan()) return false
        planRepository.save(plan)
        reschedule()
        return true
    }
}

File: apps/ios/PassAlarm/Domain/UseCase/CreatePlanUseCase.swift
import Foundation

final class CreatePlanUseCase: Sendable {
    private let planRepository: AlarmPlanRepositoryProtocol
    private let checkProLimit: CheckProLimitUseCase
    private let reschedule: RescheduleNextNUseCase

    init(planRepository: AlarmPlanRepositoryProtocol,
         checkProLimit: CheckProLimitUseCase,
         reschedule: RescheduleNextNUseCase) {
        self.planRepository = planRepository
        self.checkProLimit = checkProLimit
        self.reschedule = reschedule
    }

    enum CreatePlanError: Error {
        case limitReached
    }

    func execute(_ plan: AlarmPlan) async throws {
        let canCreate = try await checkProLimit.execute()
        guard canCreate else { throw CreatePlanError.limitReached }

        try await planRepository.save(plan)
        try await reschedule.execute()
    }
}
